### DISCUSSION TITLE
Dependency injection during binding transformation process

### DISCUSSION THREAD

**Discussion by christian-babault** [activej/activej#209] [Q&A] [answered]
Hi,

I'm creating a generic proxy, using Java dynamic proxies, which I create an instance from in a specific module.
Because it needs to be generic, I do the job in the configure method of my module, and I bind the proxy.
 
The fact is I need to inject a dependency into this proxy.
How can I get an instance of a dependency in the "configure" method of my module ?

Thanks for your answers.

Chris

**Comment by eduard-vasinskyi** [comment, marked as answer]
Hi, @christian-babault

You cannot get an instance of a dependency directly inside the `configure()` method. The Injector has not been yet created so there is nowhere to get the instance from.
However, you can manually bind your proxy class, so that it uses an instance of a dependency when it is created. Like this, for example:
```java
@Override
protected void configure() {
  . . .
  bind(Proxy.class).to(dependency -> new Proxy(dependency), Dependency.class);
}
```
If that does not solve the problem, could you please provide a snippet of your module? I’ll try to help you out with that.

**Reply by christian-babault** [reply]
Hi,

Thanks for your answer, this snippet should summarize what I need to do.

My goal is to apply a proxy transformation to any eligible class (marked with an annotation, or any specific trait), and avoid declaring each proxy explicitly.

```java
package com.kapelse.activej.module;

import io.activej.inject.module.AbstractModule;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;

public class MyProxyModule extends AbstractModule {

    private Dependency dependency; // Possible solution : find a way to inject my dependency here

    @Override
    protected void configure() {
        transform(Object.class, (bindings, scope, key, binding) -> binding.mapInstance(this::addProxy));
    }

    private <T> T addProxy(T original) {
        // At this point, the original object is already initialized
        // Possible solution : find my dependency from here, but only bindings are available
        return (T) Proxy.newProxyInstance(original.getClass().getClassLoader(),
                                          original.getClass().getInterfaces(),
                                          new MyHandler<T>(original, dependency)); // MyHandler is dynamically typed here
    }

}

// As a dynamic proxy, this class can't be a dependency itself to keep the "proxy-fication" invisible and automatic for any eligible class
public class MyHandler<T> implements InvocationHandler {

    private final T original;
    private final Dependency dependency;

    public MyHandler(T original, Dependency dependency) {
        this.original = original;
        this.dependency = dependency;
    }

    @Override
    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
        Object result;
        try {
            // Do stuff before
            result = method.invoke(original, args);
            // Do stuff after
        } catch (Throwable t) {
            // Do stuff on any error
            throw t;
        } finally {
            // Do stuff whatever happens
        }
        return result;
    }

}
```

On debug mode, I see that when _addProxy_ method is called, the _original_ object has all its dependencies injected already. No way to retrieve a dependency like I would do with Spring if I injected ApplicationContext for example ?

**Reply by eduard-vasinskyi** [reply]
As we understand it, you are trying to do the following for each instance:
1) wrap it into a proxy 
2) augment it with some kind of extra dependency, which will be used in one way or another by the proxy, along with the original instance?

If so, your example could look something like this:
```java
@Override
protected void configure() {
  transform(Object.class, (bindings, scope, key, binding) ->
        binding.addDependencies(MyExtraDependency.class)
              .mapInstance(Collections.singletonList(Key.of(MyExtraDependency.class)), (dependencies, instance) -> {
                 MyExtraDependency dependency = (MyExtraDependency) dependencies[0];
                 return addProxy(instance, dependency);
              }));
}
 
private <T> T addProxy(T original, MyExtraDependency dependency) {
  return (T) Proxy.newProxyInstance(original.getClass().getClassLoader(),
        original.getClass().getInterfaces(),
        new MyHandler<>(original, dependency));
}
```


Although that would transform any `Object` that is provided. You can use a `KeyPattern` as the first parameter of the `transform()` method to transform only instances that match the pattern.

**Reply by christian-babault** [reply]
Thank you very much, this perfectly solves the dynamic proxy dependency problem I had.

But this has revealed another problem (still in transformation process, but not about the proxy itself).
I'll open another thread for that.

**Reply by christian-babault** [reply]
By the way I cannot mark your last reply as an answer... so your first reply it will be.
